components/mpas-ocean/driver/ECDA_mpaso_obs_mod.F

```diff
@@ -0,0 +1,636 @@
+        do iob=1,NUM_OCN_obs_type
+           OCN_obs_data(iob)%assimilate=.False.    
+           allocate (OCN_obs_data(iob)%required_states(n_OCN_state), stat=ierr)
+           allocate (OCN_obs_data(iob)%update_states(n_OCN_state), stat=ierr)
+           OCN_obs_data(iob)%required_states = .FALSE.
+           OCN_obs_data(iob)%update_states   = .FALSE.
+           OCN_obs_data(iob)%grid_type    = O_GRID_MODEL       ! set model grid as default
+        enddo
+
+     end subroutine ocn_obs_init
+    
+!==============================================================================
+!BOP
+! ! Rountine set_obs_type_value(obs_data,...)
+! ! set the obs configures
+ 
+    subroutine set_obs_type_value(ID,obs_name,filename,dm_name,inst_type,assimilate, required_states, update_states)
+      
+! EOP 
+       integer, intent(in)                                    ::      ID
+       character(*),optional,intent(in)                       ::      obs_name
+       character(*),optional,intent(in)                       ::      filename
+       character(*),optional,intent(in)                       ::      dm_name
+       integer, optional,intent(in)                           ::      inst_type
+       logical, optional,intent(in)                           ::      assimilate, &
+                                                                      required_states(n_OCN_state), &
+                                                                      update_states(n_OCN_state)
+                                                       
+      if(present(obs_name)) OCN_obs_data(ID)%obs_name    = obs_name
+      if(present(filename)) OCN_obs_data(ID)%filename = filename
+      if(present(dm_name)) OCN_obs_data(ID)%dm_name = dm_name
+      if(present(inst_type)) OCN_obs_data(ID)%inst_type = inst_type
+      if(present(assimilate)) OCN_obs_data(ID)%assimilate = assimilate
+      if(present(required_states)) OCN_obs_data(ID)%required_states = required_states
+      if(present(update_states)) OCN_obs_data(ID)%update_states = update_states
+
+    end subroutine set_obs_type_value
+    
+!==============================================================================
+
+!BOP
+!   !ROUTINE: ocn_obs_type_register
+!   register the ocean observation type  for DA
+!
+!   !INTERFACE:
+
+  subroutine ocn_obs_type_register()
+!EOP
+!BOC
+        implicit none
+        ! --- Local variables ---
+        character(*), parameter :: subName =   '(ocn_obs_type_register) '
+        character(len = filename_len) :: nml_inpfile = 'ocean_da.namelist'
+        integer :: i,j,k,iob,iunit,rc,ierr
+        logical :: exists
+        logical :: obs_avalaible(max_OCN_obs_type)=.False.
+        character(len = type_name_len) ::  assim_obs_types(max_OCN_obs_type) = 'null'
+        namelist /obs_kind_nml/assim_obs_types
+        
+        NUM_OCN_obs_type=0
+        ! -read namelist for observation type        
+        inquire(file=trim(nml_inpfile), exist=exists)
```

Comment:
are there issues with multiple ranks reading this one file?  I think I've mostly seen one rank read and then broadcast to all ranks.

Reply:
I had a bad experience with broadcast mixture data with different types before.  So I set each rank read its own namelist. It works fine because the nameless is short. I can revise it with the approach of one rank read and then broadcast.